src/core/reactor_backend.cc

```diff
@@ -347,14 +351,31 @@ size_t aio_general_context::flush() {
             if (r < 0 && errno != EAGAIN) {
                 on_fatal_internal_error(seastar_logger, format("aio_general_context::flush: io_submit failed with unexpected system error: {}", errno));
             }
-            // just adjust the queue and return the number of iocbs we were able to submit
-            std::move(begin, last, iocbs.get());
             break;
         }
         begin += r;
     }
-    auto nr = begin - iocbs.get();
-    last -= nr;
+
+    auto nr = begin - original_begin;
+
+    if (nr != 0 && begin == last) {
+        // If we have succesfully committed all iocbs (begin == last) then we
+        // reset the pointers. Further at this point we move iocbs from the
+        // backlog to the main `iocbs` array if there are any.
+        // We only do this once `iocbs` is fully empty to get batching behaviour
+        // and avoid degenarate cases where only one element gets submitted
+        // which would result in excessive shifting of the elements in `iocbs`
+        // array or the backlog.
+        begin = iocbs.get();
+        last = iocbs.get();
+
+        if (!iocbs_backlog.empty()) {
+            auto max_to_copy = std::min(size_t(end - begin), iocbs_backlog.size());
+            last = std::move(iocbs_backlog.begin(), iocbs_backlog.begin() + max_to_copy, begin);
+            iocbs_backlog.erase(iocbs_backlog.begin(), iocbs_backlog.begin() + max_to_copy);
+        }
+    }
+
```

Comment:
How can we clear this backlog? If none of the already-queued IOCBs don't complete (and they're completely dependent on the peer to do so), then we'll never poll for the backlog IOCBs.

Disclaimer: haven't read the rest of the patch yet.

Reply:
Alternative: reserve one iocb for epoll, and handle the overflow by using epoll_ctl_{mod/add}

Disclaimer 1: very ugly
Disclaimer 2: may not work
Disclaimer 3: just use io_uring already
Disclaimer 4: it's slower than aio

Reply:
> How can we clear this backlog? If none of the already-queued IOCBs don't complete (and they're completely dependent on the peer to do so), then we'll never poll for the backlog IOCBs.

Yes I see your point, for a system that's in a completely idle state this would be an issue. In practice though that's probably less of an issue as something always makes progress and there is timeouts etc. Seems a bit of a fundamental issue with aio and I can't think of a nice solution immediately neither? I guess io_uring works around this kind of issue by separating the submission queue from the completion queue?

So what this patch series addresses is:
 - Benny's patches: We are temporarily not completing fast enough but retrying later will work, i.e.: don't get stuck in flush
 - My patch: We have some burst of activity and run into `last < end`. Store additional events in the backlog. Given the system is "active" it will drain soon.

We have seen those two in practice but yeah they don't address your point.

> Disclaimer 3: just use io_uring already
> Disclaimer 4: it's slower than aio

This feels like the preferred solution. We have also seen worse performance with the io_uring backend. Did you folks already had a look at why uring performs worse? Is it fundamental to io_uring or something in the seastar implemenation?

Reply:
> > How can we clear this backlog? If none of the already-queued IOCBs don't complete (and they're completely dependent on the peer to do so), then we'll never poll for the backlog IOCBs.
> 
> Yes I see your point, for a system that's in a completely idle state this would be an issue. In practice though that's probably less of an issue as something always makes progress and there is timeouts etc. 

Seems workload dependent and dangerous. 

> Seems a bit of a fundamental issue with aio and I can't think of a nice solution immediately neither? I guess io_uring works around this kind of issue by separating the submission queue from the completion queue?


io_uring works around the problem by separating the submission queue from the list of in-flight iocbs. It gains a new problem, that the number of in-flight iocbs is no longer bounded, but perhaps it has an extra limiter.

(aio was designed for disk I/O where in-flight iocbs are limited anyway)

> 
> So what this patch series addresses is:
> 
> * Benny's patches: We are temporarily not completing fast enough but retrying later will work, i.e.: don't get stuck in flush

I don't understand that patch.


> * My patch: We have some burst of activity and run into `last < end`. Store additional events in the backlog. Given the system is "active" it will drain soon.


Or maybe it won't. Consider a system with many idle clients with no pings in the protocol. Eventually the idle connections dominate and the active ones get pushed into the backlog.

> 
> We have seen those two in practice but yeah they don't address your point.
> 
> > Disclaimer 3: just use io_uring already
> > Disclaimer 4: it's slower than aio
> 
> This feels like the preferred solution. We have also seen worse performance with the io_uring backend. Did you folks already had a look at why uring performs worse? Is it fundamental to io_uring or something in the seastar implemenation?


I have vague memory of io_uring generating an internal wakeup in order to queue a completion. A scheduler trace might be better than my vague memory though. I tried it again a few months ago and it reproduced with httpd.

There's plenty of improvements we can make to the implementation:
 - instead of io_uring poll + sendmsg()/recvmsg() syscalls, collapse into io_uring sendmsg/recvmsg (without poll)
 - use lots of IORING_SETUP_* flags
 - use kernel buffer selection (advanced)
 - dedicate cores for io_uring and run with IORING_SETUP_IOPOLL/IORING_SETUP_SQPOLL [1]

Lots of fun can be had.

[1] yes it repeats an earlier point

Reply:
> > 
> > * Benny's patches: We are temporarily not completing fast enough but retrying later will work, i.e.: don't get stuck in flush
> 
> I don't understand that patch.
> 

Wait, now I do. We overshoot our preallocation, so waiting just opportunistically lets something complete (which it might not).

Better to correct our internal limits so we don't overshoot.

Reply:
Thanks for the details and sorry for the slow reply. 

> Or maybe it won't. Consider a system with many idle clients with no pings in the protocol. Eventually the idle connections dominate and the active ones get pushed into the backlog.

I think those are two different problems. 

But yeah after some reflection and internal discussion I am not super happy about the changes anymore though I still think they at least address two problems that we have seen in practice.

It does become a bit of a philosophical debate though whether crashing or perf degradation is better in this case. 

> Better to correct our internal limits so we don't overshoot.

I just find the whole programming model `--max-network-iocbs` kinda odd. To do it properly you'd have to kinda track all the sockets in your app to enforce not going over the limit. In RP we do have connection limits which would probably represent the largest source of connections so we could kinda make it work (and this already kinda happens in practice).

How do you folks do it in scylla? Do you run with the default of 10k or something higher? Given this is such a fundamental structure I am wondering whether totally overshooting and just allocating 1M entries is a bad idea? I guess it would also require adjusting aio-max-nr for large shard systems.

Reply:
> Thanks for the details and sorry for the slow reply.

I'm slow to reply too, that can make for multi-year conversations.

> 
> > Or maybe it won't. Consider a system with many idle clients with no pings in the protocol. Eventually the idle connections dominate and the active ones get pushed into the backlog.
> 
> I think those are two different problems.
> 
> But yeah after some reflection and internal discussion I am not super happy about the changes anymore though I still think they at least address two problems that we have seen in practice.
> 

I may have lost the thread, I'm not sure how I ended up on your issue tracker. So I'm not sure what's the original issue.


> It does become a bit of a philosophical debate though whether crashing or perf degradation is better in this case.
> 
> > Better to correct our internal limits so we don't overshoot.
> 
> I just find the whole programming model `--max-network-iocbs` kinda odd. To do it properly you'd have to kinda track all the sockets in your app to enforce not going over the limit. In RP we do have connection limits which would probably represent the largest source of connections so we could kinda make it work (and this already kinda happens in practice).


Yes and no. Yes, it's bad to have up-front limits and instead we should let the system discover its own limits by allocating until it runs out of resources. And we have that with epoll (although: see sysctl fs.epoll.max_user_watches) and io_uring.

No: the allocations are client-controlled and so unbounded, and also happen in the kernel which is outside the Seastar memory pool and can easily kill the system. So even with io_uring we'd want some limit.


> 
> How do you folks do it in scylla? Do you run with the default of 10k or something higher?

Run with the default.

> Given this is such a fundamental structure I am wondering whether totally overshooting and just allocating 1M entries is a bad idea? I guess it would also require adjusting aio-max-nr for large shard systems.

1M iocbs * smp::count * whatever it really occupies in the kernel -> probably too much if really pushed to the limit. So you're trading an immediate crash for slow death.


One approach is to derive a socket limit from the iocb limit and refuse to create new connections (either inbound or outbound) if we don't have resources for them. This is probably good for epoll/io_uring too. Without it, a bad client can create idle connections without bound and slowly kill the system.